src/core/ext/transport/chttp2/transport/writing.c

```diff
@@ -322,6 +322,20 @@ grpc_chttp2_begin_write_result grpc_chttp2_begin_write(
                     (send_bytes == s->compressed_data_buffer->length &&
                      s->flow_controlled_buffer.length == 0 &&
                      s->fetching_send_message == NULL);
+                if (is_last_data_frame && s->stream_compression_ctx) {
+                  GPR_ASSERT(grpc_stream_compress(
```

Comment:
Actually not completely true. Even if compression fails at this point, client can still decompress all messages. I think I can simply record the error here and continue.